### PR TITLE
MODELIX-851 Add option to exclude modules by name in bulk-sync plugin

### PR DESCRIPTION
**Pull request by mhuster23**
Aside from the new feature this PR also includes a functional test for the plugin. The already existing tests should be migrated in [MODELIX-1002](https://issues.modelix.org/issue/MODELIX-1002)

## To be verified by reviewers

* [ ] Relevant public API members have been documented
* [ ] Documentation related to this PR is complete
  * [ ] Boundary conditions are documented
  * [ ] Exceptions are documented
  * [ ] Nullability is documented if used
* [ ] Touched existing code has been extended with documentation if missing
* [ ] Code is readable
* [ ] New features and fixed bugs are covered by tests

### CODE REVIEW

**Comment by odzhychko** [comment]
Could you fix all the detekt warnings by either adding a comment or supressing them when it makes more sense?

**Comment by mhuster23** [comment]
> Could you fix all the detekt warnings by either adding a comment or supressing them when it makes more sense?

Done

**Comment by odzhychko** [comment]
> > Could you fix all the detekt warnings by either adding a comment or supressing them when it makes more sense?
> 
> Done

I meant supressing them with an annoation in the code.
By that they will not bother us while coding.

File: bulk-model-sync-gradle/src/main/kotlin/org/modelix/model/sync/bulk/gradle/config/ModelSyncGradleSettings.kt [hunk 1]
```diff
@@ -76,6 +78,14 @@ data class SyncDirection(
         (includedModulePrefixes as MutableSet).add(prefix)
     }
 
+    fun excludeModule(module: String) {
```
Comment (odzhychko) [inline bulk-model-sync-gradle/src/main/kotlin/org/modelix/model/sync/bulk/gradle/config/ModelSyncGradleSettings.kt:81]: This API is used in Gradle scripts. API documentation would be useful here. The user then does not have to go to the docs to read what this does and how it interacts with included modules.

Comment (mhuster23) [reply bulk-model-sync-gradle/src/main/kotlin/org/modelix/model/sync/bulk/gradle/config/ModelSyncGradleSettings.kt:81]: I can see the benefit of having the KDoc comment but if we do that we should do it consistently for everything (bulk-sync and api-gen) and find a way to ensure it is consistent with the docs. That's out of scope for this PR in my eyes.

Comment (odzhychko) [reply bulk-model-sync-gradle/src/main/kotlin/org/modelix/model/sync/bulk/gradle/config/ModelSyncGradleSettings.kt:81]: Ok, I understand that.

Comment (languitar) [reply bulk-model-sync-gradle/src/main/kotlin/org/modelix/model/sync/bulk/gradle/config/ModelSyncGradleSettings.kt:81]: We could just start with the methods that you created or touched, i.e. [Boy scout rule](https://innolution.com/resources/glossary/boy-scout-rule).

File: bulk-model-sync-lib/src/commonMain/kotlin/org/modelix/model/sync/bulk/Util.kt [hunk 2]
```diff
@@ -25,19 +25,27 @@ import kotlin.math.max
...
 fun isModuleIncluded(
     moduleName: String,
     includedModules: Collection<String>,
     includedPrefixes: Collection<String>,
+    excludedModules: Collection<String> = emptySet(),
+    excludedPrefixes: Collection<String> = emptySet(),
 ): Boolean {
     val includedDirectly = includedModules.contains(moduleName)
     val includedByPrefix = includedPrefixes.any { prefix -> moduleName.startsWith(prefix) }
+    val included = includedDirectly || includedByPrefix
```
Comment (odzhychko) [inline bulk-model-sync-lib/src/commonMain/kotlin/org/modelix/model/sync/bulk/Util.kt:42]: You can early return `false`, if `included` is `false`.

File: bulk-model-sync-mps/src/main/kotlin/org/modelix/mps/model/sync/bulk/IncludedModulesFilter.kt [hunk 3]
```diff
@@ -19,23 +19,34 @@ package org.modelix.mps.model.sync.bulk
...
+ *  (2) do not match the included module names ([includedModules]) or prefixes ([includedModulePrefixes]) and
+ *  (3) are not excluded via [excludedModules] or [excludedModulesPrefixes].
  *
  * Note: This is currently not meant to be used standalone.
  * It should be used with other filters in a [CompositeFilter].
  */
 class IncludedModulesFilter(
     val includedModules: Collection<String>,
     val includedModulePrefixes: Collection<String>,
+    val excludedModules: Collection<String> = emptySet(),
```
Comment (odzhychko) [inline bulk-model-sync-mps/src/main/kotlin/org/modelix/mps/model/sync/bulk/IncludedModulesFilter.kt:36]: The excludes were not passed to `IncludeModulesFilter`.

```kt
val repository = getRepository()

val includedModuleNames = parseRawPropertySet(System.getProperty("modelix.mps.model.sync.bulk.input.modules"))
val includedModulePrefixes = parseRawPropertySet(System.getProperty("modelix.mps.model.sync.bulk.input.modules.prefixes"))

val includedModulesFilter = IncludedModulesFilter(includedModuleNames, includedModulePrefixes)
```

**Review by odzhychko — CHANGES_REQUESTED** [review]

File: bulk-model-sync-mps/src/main/kotlin/org/modelix/mps/model/sync/bulk/IncludedModulesFilter.kt (hunk 3 above)
Comment (mhuster23) [reply bulk-model-sync-mps/src/main/kotlin/org/modelix/mps/model/sync/bulk/IncludedModulesFilter.kt:36]: fixed

**Review by odzhychko — APPROVED** [review]

**Pull request merged**